playground/negotiator-authenticator.js

```diff
@@ -0,0 +1,172 @@
+// import Negotiator()
+// import Authenticator()
+
+const negotiatorAuthenticator = {};
+const currentUser = 2;
+
+// Helper for success and failed promises
+const asyncHelper = async (promise) => {
+    try {
+        const data = await promise;
+        return [data, undefined];
+    } catch (error) {
+        return [undefined, error];
+    }
+}
+
+// Mock of Negotiator
+const Negotiator = {
+    getTokenInstances: function async() {
+        return [
+            { id: 42, token: { ticketClass: "VIP" }, ownerAddress: null },
+            { id: 32, token: { ticketClass: "STANDARD" }, ownerAddress: 2 },
+            { id: 15, token: { ticketClass: "VIP" }, ownerAddress: 2 },
+        ];
+    }
+};
+
+// Mock of Authenticator
+const Authenticator = {
+    authenticateAddress: async function () {
+        return true;
+    },
+    findOwner: async function () {
+        return 2
+    },
+    authenticate: async function () {
+        return "Authenticated"
+    }
+};
+
+// to provide the VIP Status and Tokens
+negotiatorAuthenticator.init = async () => {
+    // 1. Get most up to date Tokens
+    const [tokens, tokensErr] = await asyncHelper(Negotiator.getTokenInstances());
+    if (tokensErr) return negotiatorAuthenticator.errorHandler("Could not resolve token instances");
+    // 2. Determine if the user has VIP tokens (for front end developers UI / UX)
+    let isVIP = false;
+    tokens.map(ticket => { if (ticket.token.ticketClass == "VIP") isVIP = true });
+    if (isVIP) document.getElementById("vip-only-section").style.display = "block";
+    // 3. Return the and tokens
+    return tokens;
+}
+
+// to handle the vip click event from the end user
+negotiatorAuthenticator.vipClickEvent = async () => {
```

Comment:
My thought is Authenticator is the library module we provide for all tokens (e.g. Ether) hence doesn't have functions specific to a token (ticket)? So this part is the code written by a web developer, not a mock library module right?

Reply:
I maybe confused here. 

"negotiatorAuthenticator.vipClickEvent" - do you mean this function should be moved to the client side html file for the web development team manage.

Reply:
I mean:

- whatever library the TokenScript team provide to the web developers, it can't contain `vipClickEvent since it's too specific to the project the web developers are working on, therefore, negotiator-authenticator.js seem to be part of the web developers' project

- However, previously it returned mock data which should be returned by TokenScript's library (negotiator to be specific).

So the line between these 2 is not very clear to me: 𝑎) the code TokenScript libraries provides (or its mock), and 𝑏) the code web developers write

Reply:
Thanks I see what you mean. 

I'll make some updates to split the logic to it's clear to see which parts of the code are for the web developer to implement and which are TokenScript.

Keeping the web developer code inside the index.html file for now, and the tokenScript code inside the negotiator-authenticator.js file.

Reply:
@colourful-land, I've moved the logic to what I understand to be in the correct place. 
The developer code is in the index.html and the library code is inside the negotiator-authenticator.js file.